### PR TITLE
Correctly handle IP when provided with --listen

### PR DESCRIPTION
**Pull request by chsterz**
* If IP cannot be used, send a message back to the Launcher
* Scan All devices for the user-specified IP, dont just take the first IP available on the first interface
* Avoids Printing bogus IPs on startup (sometimes 'Server is running on tcp:' was printed)

### CODE REVIEW

File: core/probe.cpp [hunk 1]
```diff
@@ -416,8 +416,13 @@ void Probe::delayedInit()
     // The applicationName might be translated, so let's go with the application file base name
     m_server->setKey(QFileInfo(qApp->applicationFilePath()).completeBaseName());
     m_server->setPid(qApp->applicationPid());
-    m_server->listen();
-    ProbeSettings::sendServerAddress(m_server->externalAddress());
+    bool serverStarted = m_server->listen();
+    if (serverStarted)
+    {
```
Comment (jpetersen) [inline core/probe.cpp]: no newline

File: core/remote/tcpserverdevice.cpp [hunk 2]
```diff
@@ -61,6 +61,33 @@ bool TcpServerDevice::isListening() const
     return m_server->isListening();
 }
 
+QString TcpServerDevice::bestAvailableIP(const QHostAddress address) const
+{
+    QString result;
```
Comment (jpetersen) [inline core/remote/tcpserverdevice.cpp]: rename to firstHostFound or so

File: core/remote/tcpserverdevice.cpp [hunk 3]
```diff
@@ -61,6 +61,33 @@ bool TcpServerDevice::isListening() const
...
+
+        foreach (const QNetworkAddressEntry &addrEntry, inter.addressEntries()) {
+            const QHostAddress addr = addrEntry.ip();
+
+            // Return the ip according to the listening server protocol.
+            if (addr.protocol() != m_server->serverAddress().protocol()
+                || !addr.scopeId().isEmpty())
+                continue;
+
+            if (result.isEmpty())
```
Comment (jpetersen) [inline core/remote/tcpserverdevice.cpp]: switch both if statements (do not need to store in result if we return anyways)

File: core/probe.cpp [hunk 4]
```diff
@@ -416,8 +416,13 @@ void Probe::delayedInit()
...
+        ProbeSettings::sendServerAddress(m_server->externalAddress());
+    } else {
+        ProbeSettings::sendServerLaunchError("The address is not available");
```
Comment (vkrause) [inline core/probe.cpp]: Maybe send the QTcpServer error string here, instead the hardcoded error message?

**Comment by chsterz** [comment]
[Note to self] What I should do also: 
 - Handle the standard case (no --listen provided) such that if the launcher is started, localhost is used and use 0.0.0.0 else. 
 - If 0.0.0.0 is used, list all of the interfaces with all their IPs avail as a convenience to the user.

File: launcher/core/launcher.cpp [hunk 5]
```diff
@@ -329,6 +330,13 @@ void Launcher::readyRead()
             msg >> d->serverAddress;
             break;
         }
+        case Protocol::ServerLaunchError:
+        {
+            QString reason;
+            msg >> reason;
+            QMessageBox::critical(nullptr, "GammaRay server could not start", "The server could not start: " + reason);
```
Comment (vkrause) [inline launcher/core/launcher.cpp]: This looks problematic in terms of dependencies, as the core part of the launcher code can also used without widgets even being available, e.g. think about the CLI launcher on an embedded target.

Comment (chsterz) [inline launcher/core/launcher.cpp]: had that thought too, but must have forgotten about it when it worked so nicely ;)
I'll remove

**Pull request merged**